### PR TITLE
Add NaN checker to simulations by default

### LINKED ISSUE #938: Stop simulation when output is NaN
I was wondering whether it's a good idea to stop the simulation if the output is NaN. This can be the default and could be turned of by the user if needed (although I can't think of any situation that one wants the simulation to continue running producing NaNs...). This is useful when running simulations on HPCs. For example, when the `TimeStepWizard` checks the velocity values to recompute the time-step, if it gets a `Δt=NaN` then it should save the current state and exit the simulation?

### LINKED ISSUE #1196: Kill simulation after a crash
I am trying to reproduce a simulation similar to the one in [this paper](https://journals.ametsoc.org/jpo/article/45/6/1554/12449/The-Energetics-of-Centrifugal-Instability) and when I checked the simulation after a whole day running I found it had crashed in the first hour. (Even though I was using the TimeStepWizard.) Or at least that's what I think happened judging by the output:

```
i:  13960, sim time: 11.103 hours, wall time: 42.767 minutes, Δt: 2.875 seconds, CFL: 1.00e+00
i:  13970, sim time: 11.111 hours, wall time: 42.797 minutes, Δt: 2.875 seconds, CFL: 1.00e+00
i:  13980, sim time: 11.119 hours, wall time: 42.827 minutes, Δt: 2.875 seconds, CFL: 1.00e+00
i:  13990, sim time: 11.127 hours, wall time: 42.856 minutes, Δt: 2.875 seconds, CFL: NaN
i:  14000, sim time:  NaN years, wall time: 42.885 minutes, Δt:  NaN years, CFL: NaN
i:  14010, sim time:  NaN years, wall time: 42.913 minutes, Δt:  NaN years, CFL: NaN
```

I wonder if it's possible to implement a check that kills the simulation/throws an error when it notices that it crashed. In my LES Fortran code I implemented something like that and it prevented me from wasting many core-hours simulating `NaN`s. Basically it was just something that checked if there were any `NaN`s in the domain and killed it otherwise.

Thoughts?

### PR DESCRIPTION
**Pull request by ali-ramadhan**
This PR adds a NaN checker that checks for NaNs in the `[1, 1, 1]` grid point of the velocity fields at every time step (which should be very cheap and not affect performance).

When a NaN is detected, an `ErrorException` is thrown so when running in the REPL it'll error and return control to the REPL, while if running a script from the terminal, it will terminate the Julia session (seems like the behavior we all want).

Resolves #938
Resolves #1196

### CODE REVIEW

**Comment by navidcy** [comment]
Does it checks for NaN everywhere (i.e., in all `model.velocities` and `model.tracers`)?
If a NaN is developed somewhere then surely after 1-2 time-steps there will be NaNs everywhere, no?
So it suffices to check only one field... a field that is always there regardless the 1D/2D/3D configuration?

**Comment by ali-ramadhan** [comment]
Yeah I think checking one field should work, except for the edge case of bounded dimensions...

I guess the `[1, 1, 1]` grid point always exists and `model.velocities.{u,v,w}` always exist. But if z is `Bounded`, then `model.velocities.w[1, 1, 1]` will always be 0 (since it's halos are filled to 0 to impose no-normal flow).

And if the topology is `(Bounded, Bounded, Bounded)` then `u[1, 1, 1]`, `v[1, 1, 1]`, and `w[1, 1, 1]` will all be 0 even if the rest of the field values are all NaN.

Not sure of the best approach then. We could check `any(isnan, model.velocities.w)` less frequently?

**Comment by ali-ramadhan** [comment]
Alright so I decided to compromise so the NaN checker is always correct.

Now it goes back to checking for NaNs in the entire field (halos included). But to reduce the performance impact, the simulation's default NaN checker only checks once every 100 iterations.

File: src/Simulations/simulation.jl [hunk 1]
```diff
@@ -69,6 +69,9 @@ function Simulation(model; Δt,
              "recalculate the time step every iteration which can be slow."
    end
 
+   diagnostics[:nan_checker] = NaNChecker(fields=(u=model.velocities.u,),
+                                          schedule=IterationInterval(100))
```
Comment (glwagner) [inline src/Simulations/simulation.jl:73]: Maybe set this equal to the specified `iteration_interval`?

**Review by glwagner — APPROVED** [review]

**Comment by glwagner** [comment]
Does it make sense to add NaN checking to `stop_criteria` instead? It does seem like a stop criterion.

**Comment by navidcy** [comment]
I do prefer what @glwagner suggests. It's not a real "diagnostic" in the conventional way.

**Comment by ali-ramadhan** [comment]
Adding automatic NaN checking as a stop criteria is actually slightly different behavior because if a NaN just stops a simulation then the rest of the script still executes (instead of killing the script outright). However, the behavior could be changed by catching the right exception if we want different behavior upon seeing a NaN.

Sounds like the consensus is to refactor automatic NaN checking as a stop criteria (as part of resolving #1138)?

**Comment by glwagner** [comment]
I think the behavior one usually wants is to stop the simulation (not kill the script) upon NaN? I use this extensively for debugging (eg, use the visualization of the solution leading to blow up to diagnose an issue, etc).

**Comment by navidcy** [comment]
Ideally you want the script to kill, no?
(I'm thinking users submitting jobs on HPC and ideally you don't want to waste SUs on computing NaNs or having the job there halted because simulation stopped but the job wasn't killed and was waiting there for the rest of the requested job queued time. Right?

**Comment by ali-ramadhan** [comment]
If we agree that this PR is a net positive change can we merge it and agree to change it to a stop criteria as part of resolving #1138?

Would be nice to have so that users like @tomchor don't have to add it to the simulation callback and learn about CUDA scalar operations the hard way...

**Pull request merged**